File: 프로그래머스/lv3/42898. 등굣길/등굣길.py
# 각 블록으로 올 수 있는 경우는 오른쪽에서 진입 경우 + 위쪽에서 진입하는 경우
# 블록을 순회하면서 해당 블록을 진입할 수 있는 경우의 수 기록
# 마지막에 graph[n][m] 반환

def solution(m, n, puddles):
    graph = [[0] * (m+1) for _ in range(n+1)]
    
    for y, x in puddles: graph[x][y] = -1
    for j in range(1, m+1):
        if graph[1][j] == -1: break
        graph[1][j] = 1
    for i in range(1, n+1):
        if graph[i][1] == -1: break
        graph[i][1] = 1
    
    for i in range(2, n+1):
        for j in range(2, m+1):
            if graph[i][j] != -1:
                if graph[i-1][j] != -1:
                    graph[i][j] += graph[i-1][j]
                if graph[i][j-1] != -1:
                    graph[i][j] += graph[i][j-1]

    return graph[n][m] % 1000000007

File: 프로그래머스/lv3/42898. 등굣길/등굣길.java
// 왼쪽에서 접근 or 위쪽에서 접근
class Solution {
    public int solution(int m, int n, int[][] puddles) {
        int MOD = 1000000007;
        long[][] dp = new long[n][m];
        boolean[][] graph = new boolean[n][m];
        
        for (int[] puddle: puddles) {
            graph[puddle[1]-1][puddle[0]-1] = true;
        }
        
        for (int i=0;i<m;i++) {
            if (graph[0][i]) break;
            dp[0][i] = 1;
        }
        
        for (int i=0;i<n;i++) {
            if (graph[i][0]) break;
            dp[i][0] = 1;
        }
        
        for (int i=1;i<n;i++) {
            for (int j=1;j<m;j++) {
                if (graph[i][j]) continue;
                dp[i][j] = (dp[i-1][j] + dp[i][j-1]) % MOD;
            }
        }
        
        return (int)dp[n-1][m-1];
    }
}